File: internal/tui/review.go
package tui

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/Methamorphe/go-agent/internal/workspace"
)

type reviewSurface string

const (
	reviewConversation reviewSurface = "conversation"
	reviewPlan         reviewSurface = "plan"
	reviewDiff         reviewSurface = "diff"
)

type planStep struct {
	ID     string
	Title  string
	State  string
	Detail string
}

type planReview struct {
	BlockID   string
	ObjectRef string
	Title     string
	Steps     []planStep
}

type diffHunk struct {
	Header string
	Body   string
}

type diffFile struct {
	Path      string
	Status    string
	Additions int
	Deletions int
	Hunks     []diffHunk
}

type diffReview struct {
	BlockID   string
	ObjectRef string
	BaseRef   string
	SourceRef string
	TargetRef string
	Files     []diffFile
}

func (m Model) latestPlan() (planReview, bool) {
	for index := len(m.blocks) - 1; index >= 0; index-- {
		if m.blocks[index].Kind != workspace.BlockPlan {
			continue
		}
		return decodePlanBlock(m.blocks[index]), true
	}
	return planReview{}, false
}

func (m Model) latestDiff() (diffReview, bool) {
	for index := len(m.blocks) - 1; index >= 0; index-- {
		if m.blocks[index].Kind != workspace.BlockDiff {
			continue
		}
		return decodeDiffBlock(m.blocks[index]), true
	}
	return diffReview{}, false
}

func decodePlanBlock(block workspace.Block) planReview {
	view := planReview{BlockID: block.ID, ObjectRef: block.ObjectRef, Title: block.Title}
	if strings.TrimSpace(view.Title) == "" {
		view.Title = "Agent plan"
	}
	var raw any
	if len(block.Payload) > 0 && json.Unmarshal(block.Payload, &raw) == nil {
		body := unwrapObject(raw, "plan")
		if object, ok := body.(map[string]any); ok {
			if title := textField(object, "title", "goal", "objective"); title != "" {
				view.Title = title
			}
			if values, ok := sliceField(object, "steps", "items", "tasks"); ok {
				view.Steps = decodePlanSteps(values)
			}
		}
	}
	if len(view.Steps) == 0 {
		preview := strings.TrimSpace(block.Preview)
		if preview == "" {
			preview = "Plan artifact available at " + block.ObjectRef
		}
		view.Steps = []planStep{{ID: "1", Title: preview, State: "proposed"}}
	}
	return view
}

func decodePlanSteps(values []any) []planStep {
	steps := make([]planStep, 0, len(values))
	for index, value := range values {
		step := planStep{ID: fmt.Sprintf("%d", index+1), State: "proposed"}
		switch item := value.(type) {
		case string:
			step.Title = item
		case map[string]any:
			if id := textField(item, "id", "step_id", "key"); id != "" {
				step.ID = id
			}
			step.Title = textField(item, "title", "summary", "task", "name")
			step.Detail = textField(item, "detail", "description", "reason")
			if state := textField(item, "state", "status"); state != "" {
				step.State = state
			}
		}
		if strings.TrimSpace(step.Title) == "" {
			step.Title = fmt.Sprintf("Step %d", index+1)
		}
		steps = append(steps, step)
	}
	return steps
}

func decodeDiffBlock(block workspace.Block) diffReview {
	view := diffReview{BlockID: block.ID, ObjectRef: block.ObjectRef}
	var raw any
	if len(block.Payload) > 0 && json.Unmarshal(block.Payload, &raw) == nil {
		body := unwrapObject(raw, "diff")
		if object, ok := body.(map[string]any); ok {
			view.BaseRef = textField(object, "base_ref", "base", "base_identity")
			view.SourceRef = textField(object, "source_ref", "source", "source_identity")
			view.TargetRef = textField(object, "target_ref", "target", "target_identity")
			if values, ok := sliceField(object, "files", "changes"); ok {
				view.Files = decodeDiffFiles(values)
			}
		}
	}
	if len(view.Files) == 0 {
		path := block.ObjectRef
		if path == "" {
			path = block.Title
		}
		if path == "" {
			path = "workspace diff"
		}
		view.Files = []diffFile{{Path: path, Status: "changed", Hunks: []diffHunk{{Header: "projected diff", Body: block.Preview}}}}
	}
	return view
}

func decodeDiffFiles(values []any) []diffFile {
	files := make([]diffFile, 0, len(values))
	for _, value := range values {
		object, ok := value.(map[string]any)
		if !ok {
			continue
		}
		file := diffFile{
			Path: textField(object, "path", "file", "name"),
			Status: textField(object, "status", "state"),
			Additions: intField(object, "additions", "added"),
			Deletions: intField(object, "deletions", "removed"),
		}
		if file.Status == "" {
			file.Status = "changed"
		}
		if values, ok := sliceField(object, "hunks", "patches"); ok {
			for _, rawHunk := range values {
				switch hunk := rawHunk.(type) {
				case string:
					file.Hunks = append(file.Hunks, diffHunk{Header: "hunk", Body: hunk})
				case map[string]any:
					file.Hunks = append(file.Hunks, diffHunk{
						Header: textField(hunk, "header", "range", "title"),
						Body: textField(hunk, "body", "patch", "diff", "text"),
					})
				}
			}
		}
		if file.Path != "" {
			files = append(files, file)
		}
	}
	return files
}

func unwrapObject(value any, key string) any {
	object, ok := value.(map[string]any)
	if !ok {
		return value
	}
	if nested, ok := object[key]; ok {
		return nested
	}
	return value
}

func textField(object map[string]any, keys ...string) string {
	for _, key := range keys {
		if value, ok := object[key]; ok {
			switch typed := value.(type) {
			case string:
				if strings.TrimSpace(typed) != "" {
					return typed
				}
			case json.Number:
				return typed.String()
			case float64:
				return fmt.Sprintf("%.0f", typed)
			}
		}
	}
	return ""
}

func sliceField(object map[string]any, keys ...string) ([]any, bool) {
	for _, key := range keys {
		if value, ok := object[key].([]any); ok {
			return value, true
		}
	}
	return nil, false
}

func intField(object map[string]any, keys ...string) int {
	for _, key := range keys {
		switch value := object[key].(type) {
		case float64:
			return int(value)
		case json.Number:
			parsed, _ := value.Int64()
			return int(parsed)
		}
	}
	return 0
}

func reviewReference(kind, blockID, objectRef string) string {
	ref := blockID
	if objectRef != "" {
		ref = objectRef
	}
	if ref == "" {
		ref = "latest"
	}
	return kind + ":" + ref
}

File: cmd/go-agent-gui/frontend/src/review.ts
import type { Block } from "./types";

export interface PlanStep {
  id: string;
  title: string;
  state: string;
  detail: string;
}

export interface PlanReview {
  blockID: string;
  objectRef: string;
  title: string;
  steps: PlanStep[];
}

export interface DiffHunk { header: string; body: string }
export interface DiffFile { path: string; status: string; additions: number; deletions: number; hunks: DiffHunk[] }
export interface DiffReview {
  blockID: string;
  objectRef: string;
  baseRef: string;
  sourceRef: string;
  targetRef: string;
  files: DiffFile[];
}

function objectPayload(payload: unknown): Record<string, unknown> | null {
  if (payload && typeof payload === "object" && !Array.isArray(payload)) return payload as Record<string, unknown>;
  if (typeof payload !== "string") return null;
  try {
    const decoded = JSON.parse(payload) as unknown;
    return decoded && typeof decoded === "object" && !Array.isArray(decoded) ? decoded as Record<string, unknown> : null;
  } catch {
    return null;
  }
}

function unwrap(payload: unknown, key: string): Record<string, unknown> | null {
  const root = objectPayload(payload);
  if (!root) return null;
  const nested = root[key];
  if (nested && typeof nested === "object" && !Array.isArray(nested)) return nested as Record<string, unknown>;
  return root;
}

function text(source: Record<string, unknown>, ...keys: string[]): string {
  for (const key of keys) {
    const value = source[key];
    if (typeof value === "string" && value.trim()) return value.trim();
    if (typeof value === "number") return String(value);
  }
  return "";
}

function number(source: Record<string, unknown>, ...keys: string[]): number {
  for (const key of keys) {
    const value = source[key];
    if (typeof value === "number" && Number.isFinite(value)) return value;
  }
  return 0;
}

export function decodePlan(block: Block): PlanReview {
  const root = unwrap(block.payload, "plan");
  const title = root ? text(root, "title", "goal", "objective") : "";
  const rawSteps = root ? (root.steps ?? root.items ?? root.tasks) : null;
  const steps: PlanStep[] = [];
  if (Array.isArray(rawSteps)) {
    rawSteps.forEach((entry, index) => {
      if (typeof entry === "string") {
        steps.push({ id: String(index + 1), title: entry, state: "proposed", detail: "" });
        return;
      }
      if (!entry || typeof entry !== "object" || Array.isArray(entry)) return;
      const item = entry as Record<string, unknown>;
      steps.push({
        id: text(item, "id", "step_id", "key") || String(index + 1),
        title: text(item, "title", "summary", "task", "name") || `Step ${index + 1}`,
        state: text(item, "state", "status") || "proposed",
        detail: text(item, "detail", "description", "reason"),
      });
    });
  }
  if (steps.length === 0) {
    steps.push({ id: "1", title: block.preview || block.object_ref || "Plan available", state: "proposed", detail: "" });
  }
  return { blockID: block.id, objectRef: block.object_ref ?? "", title: title || block.title || "Agent plan", steps };
}

export function decodeDiff(block: Block): DiffReview {
  const root = unwrap(block.payload, "diff");
  const files: DiffFile[] = [];
  const rawFiles = root ? (root.files ?? root.changes) : null;
  if (Array.isArray(rawFiles)) {
    for (const entry of rawFiles) {
      if (!entry || typeof entry !== "object" || Array.isArray(entry)) continue;
      const item = entry as Record<string, unknown>;
      const hunks: DiffHunk[] = [];
      const rawHunks = item.hunks ?? item.patches;
      if (Array.isArray(rawHunks)) {
        for (const raw of rawHunks) {
          if (typeof raw === "string") hunks.push({ header: "hunk", body: raw });
          else if (raw && typeof raw === "object" && !Array.isArray(raw)) {
            const hunk = raw as Record<string, unknown>;
            hunks.push({ header: text(hunk, "header", "range", "title") || "hunk", body: text(hunk, "body", "patch", "diff", "text") });
          }
        }
      }
      const path = text(item, "path", "file", "name");
      if (path) files.push({ path, status: text(item, "status", "state") || "changed", additions: number(item, "additions", "added"), deletions: number(item, "deletions", "removed"), hunks });
    }
  }
  if (files.length === 0) {
    files.push({ path: block.object_ref || block.title || "workspace diff", status: "changed", additions: 0, deletions: 0, hunks: [{ header: "projected diff", body: block.preview ?? "" }] });
  }
  return {
    blockID: block.id,
    objectRef: block.object_ref ?? "",
    baseRef: root ? text(root, "base_ref", "base", "base_identity") : "",
    sourceRef: root ? text(root, "source_ref", "source", "source_identity") : "",
    targetRef: root ? text(root, "target_ref", "target", "target_identity") : "",
    files,
  };
}

export function reviewReference(kind: "plan" | "diff", block: Block): string {
  return `${kind}:${block.object_ref || block.id || "latest"}`;
}
